File: 백준/Gold/2504. 괄호의 값/괄호의 값.cc
#include <iostream>
#include <stack>

using namespace std;

int main(void)
{
	ios::sync_with_stdio(0);
	cin.tie(0);
	int flag = 0;
	int res = 0;
	int tmp;
	int tmp2;
	char str[31];
	stack<char> s;
	stack<int> num;
	cin >> str;
	for (int i=0; str[i] != 0; i++) {
		if (str[i] == '(' || str[i] == '[') {
			tmp = 1;
			s.push(str[i]);
		} else {
			tmp2 = 0;
			while (!s.empty() && s.top() == '.') {
				tmp2 += num.top();
				s.pop();
				num.pop();
			}
			if (s.empty()) {
				flag = 1;
				break;
			}
			if (tmp2 > 0)
				tmp = tmp2;
			if (str[i] == ')' && s.top() == '(')
				tmp *= 2;
			else if (str[i] == ']' && s.top() == '[')
				tmp *= 3;
			else {
				flag = 1;
				break;
			}
			num.push(tmp);
			s.pop();
			s.push('.');
		}
	}
	while (!s.empty() && s.top() == '.') {
		res += num.top();
		num.pop();
		s.pop();
	}
	if (flag || !s.empty())
		cout << 0;
	else
		cout << res;
	return 0;
}

File: 백준/Gold/2504. 괄호의 값/괄호의 값.py
import sys

input = sys.stdin.readline

def main():
	brackets = input().strip()
	st_brackets = []
	st_numbers = []
	flag = 0
	for bracket in brackets:
		if bracket == '(' or bracket == '[':
			tmp = 1
			st_brackets.append(bracket)
		else:
			tmp2 = 0
			while (len(st_brackets) != 0 and st_brackets[-1] == '.'):
				tmp2 += st_numbers[-1]
				st_brackets.pop()
				st_numbers.pop()
			if len(st_brackets) == 0:
				flag = 1
				break
			if tmp2 > 0:
				tmp = tmp2
			if bracket == ')' and st_brackets[-1] == '(':
				tmp *= 2
			elif bracket == ']' and st_brackets[-1] == '[':
				tmp *= 3
			else:
				flag = 1
				break
			st_numbers.append(tmp)
			st_brackets.pop()
			st_brackets.append('.')
	res = 0
	while (len(st_brackets) and st_brackets[-1] == '.'):
		res += st_numbers[-1]
		st_brackets.pop()
		st_numbers.pop()
	if (flag or len(st_brackets)):
		print(0)
	else:
		print(res)
if __name__ == "__main__":
	main()